[Feat] Enrollment 구현 및 테스트 케이스

JinYong833 opened the pull request:
## 📋 작업 내용
Entity
- Enrollment : 수강의 테이블에 필요한 필드 값 정의
- 객체가 생성될 때 default로 삽입될 값 @prepersist로 정의
DTO
- EnrollmentStatusRequest : userId, courseId, status Not null 체크
- EnrollmentProgressRequest : userId, courseId, progress 값 Not null 체크
- EnrollmentSearchRequest : userId, status Not null 체크
- EnrollmentRequest : Create만 담당하는 Request, userId, courseId 체크
- EnrollmentResponse : 테스트 결과 값 응답
Repository
- 단일 강좌 조회 optional로 null 확인 : findByUserIdAndCourseId
- userId로 수강중인 강좌 리스트 확인 : findByUserIdAndStatus
- orderNum 자동 증가를 위한 메서드 : findTopByUserIdOrderByOrderNumDesc
Service
- 수강 CRUD 비즈니스 로직 구현
TEST
- request.http : 서비스 동작 검증

## 개선 예정
- [ ] 클래스 및 구현체 네이밍 확인 및 개선 작업
- [ ] 테스트 케이스 보강

## TODO
- view 보강
- progress 초안 작업 및 팀장님에게 공유

JinYong833 committed:
Entity 패키지에 있는
Progress, Status 의 enum과
Enrollment Entity 입니다.

JinYong833 committed:
수강의 Dto입니다.
데이터 정합성을 확인하는 기능을 넣어봤습니다.
1. 불러오는 쪽에서 @Valid 사용 가능?
2. 작동 테스트 실시 해야함

JinYong833 committed:
수강의 컨트롤러 입니다.
1. 전부 Post로 동작하게 구현하였는데 논의가 필요합니다.

JinYong833 committed:
Repository 입니다.
기본 기능 외 2개 추가 되었습니다.

JinYong833 committed:
Service 클래스입니다.
1. transactional 어노테이션의 사용이 적합한지 확인 필요함
2. @Valid 의 사용이 적합한지 확인 필요함

JinYong833 committed:
readme 내용 추가 예정

JinYong833 committed:
기본 설정

JinYong833 committed:
강좌의 orderNum 필드 추가함

JinYong833 committed:
1. 불필요한 생성자 제거
2. delete 반환값 response로 수정

JinYong833 committed:
application 제거

woongblack commented on src/main/java/com/wanted/growthmate/enrollment/service/EnrollmentService.java:
```diff
@@ -0,0 +1,102 @@
+    public List<Enrollment> findEnrollmentByUserId(@Valid EnrollmentSearchRequest request) {
+        return enrollmentRepository.findByUserId(request.getUserId());
+    }
+
```
아래처럼 익셉션 추가해주시면 좋을꺼같아요!

JinYong833 replied:
감사합니다. 
나중에 다른 파트도 develop에 merge 한 이후에 반영하겠습니다.

woongblack commented:
좋은점 👍
readme.md 부분 굉장히 상세하게 써주셔서 읽어보고
코드보니 이해하기 수월했던 것 같습니다.
저는 좀 간단하게 1,2,3 단계만 썼었는데 저도 좀 정리해서 더 써보겠습니다.

dawwson commented on src/main/java/com/wanted/growthmate/enrollment/controller/EnrollmentController.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    // userId, courseId로 조회 후
+    @DeleteMapping("/delete")
+    public EnrollmentResponse deleteEnrollment(@RequestBody @Valid EnrollmentStatusRequest request) {
+        enrollmentService.deleteEnrollment(request);
+        return enrollmentService.updateEnrollmentStatus(request);
+    }
```
🔄
`@DeleteMapping`을 쓰면 `DELETE /delete`로 요청을 받는거죠?? 좀 더 RESTful 하게 `DELETE /enrollments/{enrollment_id} `이렇게 받으면 더 좋을 것 같다고 생각합니다~~

dawwson commented on src/main/java/com/wanted/growthmate/enrollment/controller/EnrollmentController.java:
```diff
@@ -0,0 +1,58 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+@RestController
```
🔄
재웅님 PR처럼 `/api`로 분리해도 괜찮을 것 같네용

dawwson commented on src/main/java/com/wanted/growthmate/enrollment/entity/Enrollment.java:
```diff
@@ -0,0 +1,124 @@
+    @Column(name = "enrollment_status", nullable = false)
+    private Status status = Status.ACTIVE;
+
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss")
```
🤔👍
신기한 어노테이션이군요... 조회할 때 저 포맷으로 바꿔주는 건가요?

dawwson commented on src/main/java/com/wanted/growthmate/enrollment/entity/Progress.java:
```diff
@@ -0,0 +1,6 @@
+package com.wanted.growthmate.enrollment.entity;
+
+public enum Progress {
+    TRUE,   // 진행 함
+    FALSE   // 진행 안함
+}
```
🔄
1차구현까지는 임시로 이렇게 처리하기로 했던 걸로 기억은 하고 있는데,,, 좀 더 도메인의 의미를 담는다면 DONE, NOT_DONE 뭐 이런식으로 하는 건 어떠신가요? boolean 값처럼 보여서 좀 헷갈릴 수도 있을 것 같습니다.

JinYong833 committed:
Enrollment Entity
1. orderNum 형식 int -> Long 으로 수정
2. setter 삭제

JinYong833 committed:
Controller
수정 없음

dawwson commented on src/main/java/com/wanted/growthmate/enrollment/dto/EnrollmentRequest.java:
```diff
@@ -0,0 +1,33 @@
+    @NotNull(message = "userId는 반드시 필요합니다.")
+    private Long userId;
+
+    @NotNull(message = "courseId는 반드시 필요합니다.")
```
오 이런 거 좋네요 👍

JinYong833 replied:
👍👍

JinYong833 committed:
Repository
수강 내역 조회를 무조건 순서 정렬하는 방식으로 수정

dawwson commented on src/main/java/com/wanted/growthmate/enrollment/request.http:
```diff
@@ -0,0 +1,58 @@
+### 등록
+POST http://localhost:8080
+Content-Type: application/json
+
+{
+  "userId": 3,
+  "courseId": 1018
+}
```
🔄
이 파일은 gitignore에 포함시켜주시면 좋을 것 같습니다!

JinYong833 replied:
넵 감사합니다.
반영하겠습니다.

dawwson reviewed (approved):
고생하셨습니다!

👍 좋은점
🔄 개선할 점
🤔 궁금한 점 
입니다! 

개선할 점은 반드시 반영해달라는 의미는 아닙니다.
지금 꼭 반영해야 하나?를 고민해보시고 수정해주시면 좋을 것 같습니다.

JinYong833 committed:
Service
1. Transactional 어노테이션 추가
2. status 변경에 따른 orderNum 수정 로직 추가
3. orderNum 수정 메서드 추가

JinYong833 committed:
orderNum 형식 수정 및 setter 삭제

JinYong833 committed:
테스트.http 추가

sobee1403 commented on src/main/java/com/wanted/growthmate/enrollment/controller/EnrollmentController.java:
```diff
@@ -0,0 +1,58 @@
+
+    // 수강 목록 생성
+    @PostMapping
+    public Enrollment createEnrollment(@RequestBody @Valid EnrollmentRequest request) {
```
@Valid 에 대해 알려주세용 수정하고 커밋친게 왜 안올라왔죠?!
문서정리해서 notion에 올려주세요

sobee1403 commented on src/main/java/com/wanted/growthmate/enrollment/entity/Enrollment.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    // 데이터 저장 전 자동으로 생성일 세팅
+    @PrePersist
```
이거에 대해 다른분들은 찾아보기~~~

sobee1403 commented on src/main/java/com/wanted/growthmate/enrollment/entity/Enrollment.java:
```diff
@@ -0,0 +1,104 @@
+        return status;
+    }
+
+    public void setStatus(Status status) {
```
setter 접근제어자가 public이어야하는이유?

JinYong833 replied:
생각해보니 말씀주신대로 entity의 setter가 public이면 안될 것 같습니다.
추후 반영하겠습니다.

sobee1403 commented on src/main/java/com/wanted/growthmate/enrollment/service/EnrollmentService.java:
```diff
@@ -0,0 +1,161 @@
+//    }
+
+    // 수강 목록 생성 (강좌 조회 및 결제와 동시에 수강 목록 생성)
+    @Transactional
```
공부해보기~~!

JinYong833 replied:
감사합니다!

sobee1403 reviewed (commented):
전반적으로 깔끔하고 좋습니다

JinYong833 committed:
view 추가
